packages/lodestar/src/chain/regen/regen.ts

```diff
@@ -0,0 +1,167 @@
+    const blockEpoch = computeEpochAtSlot(this.config, block.slot);
+    const isCheckpointBlock = block.slot % this.config.params.SLOTS_PER_EPOCH === 0;
+    if (parentEpoch < blockEpoch && !isCheckpointBlock) {
+      // If the requested state crosses an epoch boundary and the block isn't a checkpoint block
+      // then we may use the checkpoint state before the block. This may save us at least one epoch transition.
+      return this.getCheckpointState({root: block.parentRoot, epoch: blockEpoch});
+    } else if (parentEpoch < blockEpoch - 1) {
+      // If there's more than one epoch to pre-process (but the block is a checkpoint block)
+      // get the checkpoint state as close as possible
+      return this.getCheckpointState({root: block.parentRoot, epoch: blockEpoch - 1});
+    } else {
+      // Otherwise, get the state normally.
+      return this.getState(parentBlock.stateRoot);
+    }
+  }
+
+  async getCheckpointState(cp: Checkpoint): Promise<ITreeStateContext> {
+    const checkpointStartSlot = computeStartSlotAtEpoch(this.config, cp.epoch);
+    return await this.getBlockSlotState(cp.root, checkpointStartSlot);
+  }
+
+  async getBlockSlotState(blockRoot: Root, slot: Slot): Promise<ITreeStateContext> {
+    const block = this.forkChoice.getBlock(blockRoot);
+    if (!block) {
+      throw new RegenError({
+        code: RegenErrorCode.ERR_BLOCK_NOT_IN_FORKCHOICE,
+        blockRoot,
+      });
+    }
+    if (slot < block.slot) {
+      throw new RegenError({
+        code: RegenErrorCode.ERR_SLOT_BEFORE_BLOCK_SLOT,
+        slot,
+        blockSlot: block.slot,
+      });
+    }
+    const cp = {
+      root: blockRoot,
+      epoch: computeEpochAtSlot(this.config, slot),
+    };
+    const latestCheckpointStateCtx = await this.db.checkpointStateCache.getLatest(cp);
+    if (latestCheckpointStateCtx) {
+      // If a checkpoint state exists with the given checkpoint root, it either is in requested epoch
+      // or needs to have empty slots processed until the requested epoch
+      return await processSlotsByCheckpoint(this.emitter, latestCheckpointStateCtx, slot);
+    } else {
+      // Otherwise, use the fork choice to get the stateRoot from block at the checkpoint root
+      // regenerate that state,
+      // then process empty slots until the requested epoch
+      const blockStateCtx = await this.getState(block.stateRoot);
+      return await processSlotsByCheckpoint(this.emitter, blockStateCtx, slot);
+    }
+  }
+
+  async getState(stateRoot: Root): Promise<ITreeStateContext> {
+    // Trivial case, stateCtx at stateRoot is already cached
+    const cachedStateCtx = await this.db.stateCache.get(stateRoot);
+    if (cachedStateCtx) {
+      return cachedStateCtx;
+    }
+    // Otherwise we have to use the fork choice to traverse backwards, block by block,
+    // searching the state caches
+    // then replay blocks forward to the desired stateRoot
+    const _Root = this.config.types.Root;
```

Comment:
```suggestion
    const rootType = this.config.types.Root;
```
to keep naming consistent